proposals/shared-everything-threads/Overview.md

```diff
@@ -666,6 +673,7 @@ the memory is `shared`. Likewise:
 | `array.atomic.rmw.xor <u32:ordering> <typeidx>` | 0xFE 0x6F | valid for i32 and i64 arrays. |
 | `array.atomic.rmw.xchg <u32:ordering> <typeidx>` | 0xFE 0x70 | valid for i32, i64, and <: anyref arrays. |
 | `array.atomic.rmw.cmpxchg <u32:ordering> <typeidx>` | 0xFE 0x71 | valid for i32, i64, and <: eqref arrays. |
+| `ref.shared.eq` | 0xFE 0x72 | valid for shared eqref operands |
```

Comment:
I don't think we need a separate instruction. The principal types property can be maintained by allowing a meta-variable for the sharing attribute, analogous to how it already deals with the nullability attribute.

Reply:
Can you expand on this a little bit? I thought that part of the reason we now have extra annotations on `br_cast` was to avoid "polymorphism in nullability".

Reply:
IIRC, the reason we needed to be explicit about null for casts is that we need to express how null is treated _operationally_, i.e., `ref.cast (ref eq)` is a different operation from `ref.cast (ref null eq)`, not just differently typed (insert obligatory complaint about the monolithic nature of our cast instructions here).

Parametric nullability comes up, for example, in the case of `extern.convert_any`, which has type `(REF nul ANY) -> (REF nul EXTERN)` for either choice of `nul`.

Similarly, we'd have
```
ref.eq : (REF NULL (sh EQ)) (REF NULL (sh EQ)) → I32
```
for either choice of `sh`.

Reply:
Thanks, I'll add a note about sharedness metavariables being allowed in our principal types rule and that all instructions that operate on unshared references should be allowed to work with shared references as well. `array.len` is another such instruction that I can think of.